codec.go

```diff
@@ -1,40 +1,100 @@
 package rpc
 
 import (
-	"github.com/ugorji/go/codec"
 	"io"
+
+	"github.com/ugorji/go/codec"
 )
 
 type decoder interface {
 	Decode(interface{}) error
 }
 
-type byteReadingDecoder interface {
-	decoder
-	io.ByteReader
+type encoder interface {
+	Encode(interface{}) <-chan error
 }
 
-type encoder interface {
-	Encode(interface{}) error
+const poolSize int = 10
+
+type decoderPool chan *codec.Decoder
+
+func makeDecoderPool() decoderPool {
+	p := make(decoderPool, poolSize)
+
+	for i := 0; i < poolSize; i++ {
+		p <- codec.NewDecoderBytes([]byte{}, newCodecMsgpackHandle())
+	}
+
+	return p
+}
+
+func (p decoderPool) getDecoder(bytes []byte) *codec.Decoder {
+	dec := <-p
+	dec.ResetBytes(bytes)
+	return dec
+}
+
+func (p decoderPool) returnDecoder(dec *codec.Decoder) {
+	p <- dec
+}
+
+type encoderPool chan *codec.Encoder
+
+func makeEncoderPool() encoderPool {
+	p := make(encoderPool, poolSize)
+
+	for i := 0; i < poolSize; i++ {
+		p <- codec.NewEncoderBytes(&[]byte{}, newCodecMsgpackHandle())
+	}
+
+	return p
+}
+
+func (p encoderPool) getEncoder(bytes *[]byte) *codec.Encoder {
+	enc := <-p
+	enc.ResetBytes(bytes)
+	return enc
+}
+
+func (p encoderPool) returnEncoder(enc *codec.Encoder) {
+	p <- enc
+}
+
+func newCodecMsgpackHandle() codec.Handle {
+	return &codec.MsgpackHandle{
+		WriteExt:    true,
+		RawToString: true,
+	}
+}
+
+type writeBundle struct {
+	bytes []byte
+	ch    chan error
 }
 
 type framedMsgpackEncoder struct {
-	handle   codec.Handle
-	writeCh  chan []byte
-	resultCh chan error
+	encoders encoderPool
+	writer   io.Writer
+	writeCh  chan writeBundle
+	doneCh   chan struct{}
+	closedCh chan struct{}
 }
 
-func newFramedMsgpackEncoder(writeCh chan []byte, resultCh chan error) *framedMsgpackEncoder {
-	mh := &codec.MsgpackHandle{WriteExt: true}
-	return &framedMsgpackEncoder{
-		handle:   mh,
-		writeCh:  writeCh,
-		resultCh: resultCh,
+func newFramedMsgpackEncoder(writer io.Writer) *framedMsgpackEncoder {
+	e := &framedMsgpackEncoder{
+		encoders: makeEncoderPool(),
+		writer:   writer,
+		writeCh:  make(chan writeBundle),
+		doneCh:   make(chan struct{}),
+		closedCh: make(chan struct{}),
 	}
+	go e.writerLoop()
+	return e
 }
 
 func (e *framedMsgpackEncoder) encodeToBytes(i interface{}) (v []byte, err error) {
-	enc := codec.NewEncoderBytes(&v, e.handle)
+	enc := e.encoders.getEncoder(&v)
```

Comment:
You'll note that this is only an in-memory encode from a data structure to a byte slice. I can't use the same encoder because I need to change the byte slice it's encoding to, and this ends up being simpler than adding coordination channels to the writer loop. 10 encoders was a rough guess, but I this encode is a glorified `memcpy`, so I don't anticipate them being in use for much of the execution path.